cmd/reaper/types_watcher.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/boskos/common"
+)
+
+type ResourceTypesWatcher struct {
+	config        string
+	providedTypes []string
+	mergedTypes   []string
+	l             sync.Locker
+}
+
+func NewResourceTypesWatcher(providedTypes []string, config string) (*ResourceTypesWatcher, error) {
+	rt := ResourceTypesWatcher{
+		config:        config,
+		providedTypes: providedTypes,
+		l:             &sync.Mutex{},
+	}
+
+	if config != "" {
+		// Read and parse the types for the first time
+		// In this way 'ResourceTypesWatcher.Types()' returns the correct result
+		// even if the config is never updated
+		configResources, err := parseResourceTypesFromConfig(rTypesConfig)
+		if err != nil {
+			return nil, fmt.Errorf("parse resource types: %w", err)
+		}
+		rt.mergedTypes = mergeResourceTypes(configResources, providedTypes)
+		rt.watchConfig()
+	} else {
+		rt.mergedTypes = providedTypes
+	}
+
+	return &rt, nil
+}
+
+func (rtw *ResourceTypesWatcher) watchConfig() {
+	go func() {
+		v := viper.New()
+		v.SetConfigFile(rtw.config)
+		v.SetConfigType("yaml")
+		v.OnConfigChange(func(in fsnotify.Event) {
+			// Don't block the handler on reading and parsing the new config
```

Comment:
This means that multiple updates in quick succession will race to update the internal state, and you are not guaranteed to converge to the correct state (on disk). Check out the ConfigAgent implementation.